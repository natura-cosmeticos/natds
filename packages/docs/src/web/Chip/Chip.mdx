import { Meta, Story, Preview } from '@storybook/addon-docs/blocks';
import { Chip } from '@naturacosmeticos/natds-web';

# Chip

Chips are compact elements that represent an input, attribute, or action.

### Usage

Chips allow users to enter information, make selections, filter content, or trigger actions. Chips should appear dynamically as a group of multiple interactive elements. Unlike buttons, which should be a consistent and familiar call to action, one that a user expects to appear as the same action in the same general area.

***

### Principles

***

#### Compact

Chips are compact components that represent discrete information.

#### Relevant

Chips should have a clear and helpful relationship to the content or task they represent.

#### Focused

Chips should make tasks easier to complete, or content easier to sort.

<Preview>
  <Chip label="label" />
</Preview>

<Story id="web-chip--default" />
