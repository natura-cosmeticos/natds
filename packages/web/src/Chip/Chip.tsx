import React from 'react';

interface IProps {
  labelOn: string;
  labelOff: string;
};

interface IState {
  isChecked: boolean;
};

export class Chip extends React.Component<{
  labelOn: string;
  labelOff: string;
}, IState> {
  constructor(props: IProps) {
    super(props);
    this.state = { isChecked: false };
  }

  onChange = () => {
    this.setState({ isChecked: !this.state.isChecked });
  }

  render() {
    return (
      <label>
        <input
          type="checkbox"
          checked={this.state.isChecked}
          onChange={this.onChange}
        />
        {this.state.isChecked ? this.props.labelOn : this.props.labelOff}
      </label>
    );
  }
};
