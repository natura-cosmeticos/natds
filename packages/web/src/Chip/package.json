{
  "name": "Checkbox",
  "main": "./Checkbox.js",
  "private": true
}
