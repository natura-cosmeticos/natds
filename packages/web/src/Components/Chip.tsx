import React, { FunctionComponent, forwardRef } from "react";
import { withTheme } from "@material-ui/styles";
import MaterialChip, { ChipProps } from "@material-ui/core/Chip";
import { IThemeWeb } from "Themes";
import Icon from "./Icon";

export interface IChipProps extends ChipProps {

  /**
   * @optional
   */
  theme?: IThemeWeb | unknown;
}

export const Chip: FunctionComponent<IChipProps> = forwardRef((
  props: IChipProps,
  ref: any,
) => {
  const { deleteIcon: providedDeletedIcon } = props;
  const deleteIcon = React.isValidElement(providedDeletedIcon)
    ? providedDeletedIcon
    : <Icon name="filled-action-cancel" />;

  return (
    <MaterialChip {...props} deleteIcon={deleteIcon} ref={ref} />
  );
});

export default withTheme(Chip);
