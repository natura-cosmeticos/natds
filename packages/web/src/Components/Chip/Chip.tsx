import * as React from 'react'
import MaterialChip from '@material-ui/core/Chip'
import { Icon } from '../Icon'
import { IChipProps } from './Chip.props'

/**
 * For more advanced usages, check [Material UI Chip docs](https://material-ui.com/components/chips/)
 *
 * ## Importing
 *
 * ```
 * import { Chip } from '@naturacosmeticos/natds-web';
 * ```
 *
 * @see https://material-ui.com/components/chips/
 */
export const Chip = React.forwardRef<HTMLDivElement, IChipProps>((
  props: IChipProps, ref
) => {
  const {
    deleteIcon, onDelete, size, ...otherProps
  } = props

  const iconFactory = () => {
    if (!React.isValidElement(deleteIcon)) {
      const iconSize = size === 'small' ? 'micro' : 'tiny'

      return <Icon name="filled-action-cancel" size={iconSize} />
    }

    return deleteIcon
  }

  /**
   * @see https://material-ui.com/guides/composition/#caveat-with-inlining
   */
  const validDeleteIcon = React.useMemo(iconFactory, [deleteIcon, size])

  return (
    <MaterialChip
      deleteIcon={onDelete && validDeleteIcon}
      onDelete={onDelete}
      size={size}
      {...otherProps}
      ref={ref}
    />
  )
})

Chip.displayName = 'Chip'

export default Chip
