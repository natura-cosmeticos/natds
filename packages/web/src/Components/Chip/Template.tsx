import { Story } from '@storybook/react'
import * as React from 'react'
import { IChipProps } from './Chip.props'
import { Chip } from './Chip'
import { variants } from './__fixtures__/variants'

export const Template : Story<IChipProps> = (args: IChipProps) => <Chip {...args} />

export const TemplateWithVariants : Story<IChipProps> = (args: IChipProps) => (
  <div>
    <Chip {...args} variant={variants.default} />
    {' '}
    <Chip {...args} variant={variants.outlined} />
  </div>
)

export default Template
