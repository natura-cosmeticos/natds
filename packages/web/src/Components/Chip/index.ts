import Chip from './Chip'

export { IChipProps } from './Chip.props'
export { Chip }
export default Chip
