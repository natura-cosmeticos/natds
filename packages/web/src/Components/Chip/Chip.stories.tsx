/* eslint-disable max-lines */
import * as React from 'react'
import { Meta, Story } from '@storybook/react'
import { Chip } from './Chip'
import { IChipProps } from './Chip.props'
import { colors } from './__fixtures__/colors'
import { sizes } from './__fixtures__/sizes'
import { variants } from './__fixtures__/variants'
import { Icon } from '../Icon'
import { Avatar } from '../Avatar'
import { Letter, Playground as AvatarPlayground } from '../Avatar/Avatar.stories'
import { Template, TemplateWithVariants } from './Template'
import { TinyIcon } from '../Icon/Icon.stories'

export default {
  component: Chip,
  subcomponents: { Icon },
  title: 'Components/Chip'
} as Meta

const customDeleteIcon = <Icon {...TinyIcon.args} name="filled-action-delete" />

export const Playground : Story<IChipProps> = Template.bind({})
Playground.args = {
  clickable: true,
  color: colors.default,
  disabled: false,
  label: 'Chip component label',
  size: sizes.medium,
  variant: variants.default
}

export const Standard : Story<IChipProps> = TemplateWithVariants.bind({})
Standard.args = {
  color: colors.primary,
  label: 'Label',
  size: sizes.medium,
  variant: variants.default
}

export const Clickable : Story<IChipProps> = TemplateWithVariants.bind({})
Clickable.args = {
  ...Standard.args,
  clickable: true
}

export const DefaultVariant : Story<IChipProps> = Template.bind({})
DefaultVariant.args = {
  ...Clickable.args,
  variant: variants.default
}

export const OutlinedVariant : Story<IChipProps> = Template.bind({})
OutlinedVariant.args = {
  ...Clickable.args,
  variant: variants.outlined
}

export const Disabled : Story<IChipProps> = TemplateWithVariants.bind({})
Disabled.args = {
  ...Standard.args,
  disabled: true
}

export const WithoutDeleteIcon : Story<IChipProps> = TemplateWithVariants.bind({})
WithoutDeleteIcon.args = {
  ...Standard.args,
  onDelete: null
}

export const WithCustomDeleteIcon : Story<IChipProps> = TemplateWithVariants.bind({})
WithCustomDeleteIcon.args = {
  ...Standard.args,
  deleteIcon: customDeleteIcon
}

export const Primary : Story<IChipProps> = TemplateWithVariants.bind({})
Primary.args = {
  ...Clickable.args,
  color: colors.primary
}

export const Secondary : Story<IChipProps> = TemplateWithVariants.bind({})
Secondary.args = {
  ...Clickable.args,
  color: colors.secondary
}

export const Default : Story<IChipProps> = TemplateWithVariants.bind({})
Default.args = {
  ...Clickable.args,
  color: colors.default
}

export const Small : Story<IChipProps> = TemplateWithVariants.bind({})
Small.args = {
  ...Clickable.args,
  size: sizes.small
}

export const Medium : Story<IChipProps> = TemplateWithVariants.bind({})
Medium.args = {
  ...Clickable.args,
  size: sizes.medium
}

export const WithAvatar : Story<IChipProps> = TemplateWithVariants.bind({})
WithAvatar.args = {
  ...Standard.args,
  avatar: <Avatar {...AvatarPlayground.args} />
}

export const WithLetterAvatar : Story<IChipProps> = TemplateWithVariants.bind({})
WithLetterAvatar.args = {
  ...Standard.args,
  avatar: <Avatar {...Letter.args} />
}

export const WithTinyIcon : Story<IChipProps> = TemplateWithVariants.bind({})
WithTinyIcon.args = {
  ...Standard.args,
  icon: <Icon {...TinyIcon.args} name="filled-default-mockup" />
}
