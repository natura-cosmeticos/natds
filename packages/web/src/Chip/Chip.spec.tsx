
import React from 'react';
import { shallow } from 'enzyme'
import { Chip } from './Chip'

test('Chip changes the text after click', () => {
  const chip = shallow(<Chip labelOn="On" labelOff="Off" />);

  expect(chip.text()).toEqual('Off')
  chip.find('input').simulate('change')
  expect(chip.text()).toEqual('On')

  expect(chip).toMatchSnapshot()
})
