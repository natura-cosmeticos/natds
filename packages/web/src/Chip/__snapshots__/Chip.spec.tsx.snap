// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip changes the text after click 1`] = `
<label>
  <input
    checked={true}
    onChange={[Function]}
    type="checkbox"
  />
  On
</label>
`;
